File: apps/api/prisma/seed.ts
import {
  BALevel,
  BAStatus,
  BookingPriority,
  BookingStatus,
  PrismaClient,
  SkillTagGroup,
  SkillTagStatus,
  UserRole
} from '@prisma/client';
import { PrismaPg } from '@prisma/adapter-pg';

const localDevDatabaseUrl =
  'postgresql://ba_bazaar:change_me@localhost:5433/ba_bazaar?schema=public';

process.env.DATABASE_URL ??= localDevDatabaseUrl;

const prisma = new PrismaClient({
  adapter: new PrismaPg({
    connectionString: process.env.DATABASE_URL
  })
});

const date = (value: string) => new Date(`${value}T00:00:00.000Z`);

async function resetDatabase() {
  await prisma.auditLog.deleteMany();
  await prisma.notification.deleteMany();
  await prisma.privateNote.deleteMany();
  await prisma.booking.deleteMany();
  await prisma.bASkillTag.deleteMany();
  await prisma.skillTag.deleteMany();
  await prisma.project.deleteMany();
  await prisma.bAProfile.deleteMany();
  await prisma.user.deleteMany();
}

async function main() {
  await resetDatabase();

  const manager = await prisma.user.create({
    data: {
      full_name: 'Mai Lan Anh',
      email: 'lan.anh.manager@ba-bazaar.local',
      role: UserRole.BA_MANAGER,
      avatar_url: 'https://api.dicebear.com/9.x/initials/svg?seed=Mai%20Lan%20Anh'
    }
  });

  const pmUsers = await Promise.all(
    ['Minh Tran', 'Hoa Nguyen', 'Quang Pham', 'Linh Do', 'Khanh Vo'].map(
      (name, index) =>
        prisma.user.create({
          data: {
            full_name: name,
            email: `pm${index + 1}@ba-bazaar.local`,
            role: UserRole.PM_PO,
            avatar_url: `https://api.dicebear.com/9.x/initials/svg?seed=${encodeURIComponent(name)}`
          }
        })
    )
  );

  const tagInputs = [
    ['Fintech', SkillTagGroup.DOMAIN],
    ['E-commerce', SkillTagGroup.DOMAIN],
    ['Logistics', SkillTagGroup.DOMAIN],
    ['HR Tech', SkillTagGroup.DOMAIN],
    ['CRM', SkillTagGroup.DOMAIN],
    ['BPMN', SkillTagGroup.ANALYSIS_SKILL],
    ['User Story Mapping', SkillTagGroup.ANALYSIS_SKILL],
    ['Data Analysis', SkillTagGroup.ANALYSIS_SKILL],
    ['API Specification', SkillTagGroup.ANALYSIS_SKILL],
    ['Stakeholder Workshop', SkillTagGroup.ANALYSIS_SKILL]
  ] as const;

  const tags = await Promise.all(
    tagInputs.map(([name, group]) =>
      prisma.skillTag.create({
        data: {
          name,
          group,
          status: SkillTagStatus.ACTIVE
        }
      })
    )
  );
  const tagByName = new Map(tags.map((tag) => [tag.name, tag]));

  const projects = await Promise.all(
    [
      ['Payment Refund Flow', '#2563EB', 'Refund and reconciliation workflow'],
      ['Mobile Onboarding', '#16A34A', 'Digital onboarding for mobile users'],
      ['CRM Revamp', '#7C3AED', 'Internal CRM modernization'],
      ['Logistics Tracking', '#F97316', 'Shipment tracking and exception flows'],
      ['BI Dashboard', '#0F766E', 'Executive utilization and delivery dashboard'],
      ['HR Approval Workflow', '#DB2777', 'People operation request approvals']
    ].map(([name, color, description]) =>
      prisma.project.create({
        data: {
          name,
          color,
          description
        }
      })
    )
  );

  const baInputs = [
    ['Pham Ngoc Chi', BALevel.SENIOR, BAStatus.ACTIVE, ['Fintech', 'BPMN', 'API Specification']],
    ['Do Anh Dung', BALevel.MIDDLE, BAStatus.ACTIVE, ['Logistics', 'Data Analysis']],
    ['Nguyen Bao An', BALevel.JUNIOR, BAStatus.ACTIVE, ['HR Tech', 'User Story Mapping']],
    ['Le Dang Khoa', BALevel.MIDDLE, BAStatus.ACTIVE, ['CRM', 'Stakeholder Workshop']],
    ['Bui Phuong Thao', BALevel.SENIOR, BAStatus.ACTIVE, ['CRM', 'Data Analysis']],
    ['Hoang Minh Chau', BALevel.LEAD, BAStatus.ACTIVE, ['Fintech', 'CRM', 'BPMN']],
    ['Tran Gia Huy', BALevel.MIDDLE, BAStatus.ACTIVE, ['E-commerce', 'API Specification']],
    ['Vo Thanh Tam', BALevel.SENIOR, BAStatus.ACTIVE, ['Logistics', 'BPMN']],
    ['Dang Thu Ha', BALevel.JUNIOR, BAStatus.ACTIVE, ['CRM', 'User Story Mapping']],
    ['Nguyen Mai Linh', BALevel.MIDDLE, BAStatus.ACTIVE, ['Fintech', 'Data Analysis']],
    ['Pham Quoc Bao', BALevel.SENIOR, BAStatus.ACTIVE, ['E-commerce', 'Stakeholder Workshop']],
    ['Do Minh Tue', BALevel.MIDDLE, BAStatus.ACTIVE, ['HR Tech', 'BPMN']],
    ['Le Hoai Nam', BALevel.LEAD, BAStatus.ON_LEAVE, ['CRM', 'API Specification']],
    ['Vu Nhat Vy', BALevel.SENIOR, BAStatus.ON_LEAVE, ['Fintech', 'Data Analysis']],
    ['Tran Thai Son', BALevel.MIDDLE, BAStatus.RESIGNED, ['Logistics', 'BPMN']]
  ] as const;

  const bas = [];
  for (const [index, [name, level, status, tagNames]] of baInputs.entries()) {
    const user = await prisma.user.create({
      data: {
        full_name: name,
        email: `ba${index + 1}@ba-bazaar.local`,
        role: UserRole.BA,
        avatar_url: `https://api.dicebear.com/9.x/initials/svg?seed=${encodeURIComponent(name)}`
      }
    });

    const ba = await prisma.bAProfile.create({
      data: {
        user_id: user.id,
        full_name: name,
        email: user.email,
        phone: `090${String(index + 1).padStart(7, '0')}`,
        level,
        joined_date: date(`202${Math.min(index % 5, 5)}-0${(index % 8) + 1}-15`),
        avatar_url: user.avatar_url,
        status,
        status_reason:
          status === BAStatus.ON_LEAVE
            ? 'Temporary leave'
            : status === BAStatus.RESIGNED
              ? 'Historical profile retained'
              : null,
        status_changed_at: status === BAStatus.ACTIVE ? null : date('2026-05-20')
      }
    });

    for (const tagName of tagNames) {
      const tag = tagByName.get(tagName);
      if (!tag) {
        continue;
      }

      await prisma.bASkillTag.create({
        data: {
          ba_id: ba.id,
          tag_id: tag.id,
          assigned_by: manager.id
        }
      });
    }

    bas.push(ba);
  }

  const projectByName = new Map(projects.map((project) => [project.name, project]));

  const bookingInputs = [
    [0, 'Payment Refund Flow', 0, 'Refund analysis sprint', '2026-06-01', '2026-06-05', 50, BookingStatus.APPROVED, BookingPriority.HIGH],
    [0, 'CRM Revamp', 1, 'Pending CRM dependency mapping', '2026-06-03', '2026-06-06', 50, BookingStatus.PENDING, BookingPriority.MEDIUM],
    [1, 'Logistics Tracking', 1, 'Shipment exception flow', '2026-06-01', '2026-06-06', 100, BookingStatus.APPROVED, BookingPriority.HIGH],
    [2, 'HR Approval Workflow', 2, 'HR request rules', '2026-06-03', '2026-06-04', 100, BookingStatus.APPROVED, BookingPriority.LOW],
    [4, 'BI Dashboard', 3, 'Dashboard requirements', '2026-06-01', '2026-06-05', 50, BookingStatus.APPROVED, BookingPriority.MEDIUM],
    [4, 'Internal Portal', 4, 'Portal request risk', '2026-06-01', '2026-06-06', 100, BookingStatus.PENDING, BookingPriority.URGENT],
    [5, 'CRM Revamp', 0, 'Lead BA CRM review', '2026-06-02', '2026-06-07', 100, BookingStatus.IN_PROGRESS, BookingPriority.HIGH],
    [6, 'Mobile Onboarding', 1, 'Onboarding discovery', '2026-06-09', '2026-06-13', 50, BookingStatus.PENDING, BookingPriority.MEDIUM],
    [7, 'Logistics Tracking', 2, 'Carrier integration analysis', '2026-06-10', '2026-06-14', 50, BookingStatus.APPROVED, BookingPriority.HIGH],
    [8, 'CRM Revamp', 3, 'Customer profile grooming', '2026-06-09', '2026-06-10', 100, BookingStatus.REJECTED, BookingPriority.LOW],
    [10, 'BI Dashboard', 1, 'Completed onboarding metric wrap-up', '2026-06-02', '2026-06-03', 50, BookingStatus.COMPLETED, BookingPriority.MEDIUM],
    [8, 'Internal Portal', 1, 'Portal backlog walkthrough', '2026-05-21', '2026-05-23', 50, BookingStatus.COMPLETED, BookingPriority.MEDIUM],
    [8, 'BI Dashboard', 4, 'CRM metrics mapping', '2026-05-12', '2026-05-16', 50, BookingStatus.COMPLETED, BookingPriority.HIGH],
    [8, 'Payment Refund Flow', 0, 'Refund contact-center scenarios', '2026-04-22', '2026-04-25', 50, BookingStatus.COMPLETED, BookingPriority.MEDIUM],
    [8, 'Mobile Onboarding', 2, 'Onboarding support fallback', '2026-04-08', '2026-04-10', 50, BookingStatus.CANCELLED, BookingPriority.LOW],
    [8, 'HR Approval Workflow', 1, 'Policy intake notes', '2026-03-17', '2026-03-20', 50, BookingStatus.COMPLETED, BookingPriority.LOW],
    [8, 'CRM Revamp', 0, 'Legacy contact merge review', '2026-02-10', '2026-02-13', 50, BookingStatus.COMPLETED, BookingPriority.MEDIUM],
    [8, 'Logistics Tracking', 2, 'Returned-order exception analysis', '2026-01-19', '2026-01-23', 100, BookingStatus.COMPLETED, BookingPriority.HIGH],
    [8, 'Payment Refund Flow', 4, 'Refund SLA gap mapping', '2025-12-08', '2025-12-12', 50, BookingStatus.COMPLETED, BookingPriority.MEDIUM],
    [8, 'BI Dashboard', 3, 'Retention metrics workshop', '2025-11-17', '2025-11-21', 100, BookingStatus.COMPLETED, BookingPriority.HIGH],
    [8, 'Internal Portal', 1, 'Service request taxonomy cleanup', '2025-10-06', '2025-10-10', 50, BookingStatus.COMPLETED, BookingPriority.LOW],
    [8, 'CRM Revamp', 0, 'Lead intake form simplification', '2025-09-15', '2025-09-19', 100, BookingStatus.COMPLETED, BookingPriority.HIGH],
    [8, 'HR Approval Workflow', 4, 'Escalation ladder discovery', '2025-08-11', '2025-08-15', 50, BookingStatus.COMPLETED, BookingPriority.MEDIUM],
    [8, 'Mobile Onboarding', 2, 'Activation funnel notes', '2025-07-07', '2025-07-11', 100, BookingStatus.COMPLETED, BookingPriority.MEDIUM],
    [8, 'Payment Refund Flow', 3, 'Refund reasons catalog', '2025-06-09', '2025-06-13', 50, BookingStatus.COMPLETED, BookingPriority.LOW],
    [8, 'BI Dashboard', 1, 'Customer health score definition', '2025-05-12', '2025-05-16', 100, BookingStatus.COMPLETED, BookingPriority.HIGH],
    [8, 'CRM Revamp', 4, 'Duplicate profile cleanup rules', '2025-04-14', '2025-04-18', 50, BookingStatus.COMPLETED, BookingPriority.MEDIUM],
    [9, 'Payment Refund Flow', 4, 'Refund report review', '2026-06-11', '2026-06-15', 50, BookingStatus.APPROVED, BookingPriority.MEDIUM],
    [10, 'Mobile Onboarding', 0, 'KYC rules', '2026-06-16', '2026-06-20', 100, BookingStatus.PENDING, BookingPriority.URGENT],
    [11, 'HR Approval Workflow', 1, 'Approval matrix', '2026-06-16', '2026-06-18', 50, BookingStatus.APPROVED, BookingPriority.MEDIUM],
    [0, 'Payment Refund Flow', 2, 'Completed refund baseline', '2026-05-04', '2026-05-08', 50, BookingStatus.COMPLETED, BookingPriority.MEDIUM],
    [1, 'Logistics Tracking', 3, 'Cancelled logistics support', '2026-05-11', '2026-05-13', 50, BookingStatus.CANCELLED, BookingPriority.LOW],
    [2, 'HR Approval Workflow', 4, 'Completed HR interviews', '2026-05-18', '2026-05-22', 100, BookingStatus.COMPLETED, BookingPriority.HIGH],
    [3, 'CRM Revamp', 0, 'Free-capacity pending demo', '2026-06-04', '2026-06-10', 50, BookingStatus.PENDING, BookingPriority.HIGH],
    [5, 'BI Dashboard', 1, 'Approved full capacity block', '2026-06-10', '2026-06-14', 100, BookingStatus.APPROVED, BookingPriority.URGENT],
    [5, 'Mobile Onboarding', 2, 'Should be blocked if approved', '2026-06-11', '2026-06-12', 50, BookingStatus.PENDING, BookingPriority.HIGH],
    [12, 'CRM Revamp', 3, 'Historical on-leave booking', '2026-05-01', '2026-05-03', 50, BookingStatus.COMPLETED, BookingPriority.LOW],
    [14, 'Logistics Tracking', 4, 'Historical resigned BA booking', '2026-04-01', '2026-04-05', 100, BookingStatus.COMPLETED, BookingPriority.MEDIUM]
  ] as const;

  const createdBookings = [];
  for (const [
    baIndex,
    projectName,
    requesterIndex,
    title,
    startDate,
    endDate,
    capacityPercent,
    status,
    priority
  ] of bookingInputs) {
    const project = projectByName.get(projectName) ?? projects[0];
    const approved = [
      BookingStatus.APPROVED,
      BookingStatus.IN_PROGRESS,
      BookingStatus.COMPLETED
    ].includes(status);
    const rejected = status === BookingStatus.REJECTED;
    const cancelled = status === BookingStatus.CANCELLED;

    const booking = await prisma.booking.create({
      data: {
        ba_id: bas[baIndex].id,
        project_id: project.id,
        requester_id: pmUsers[requesterIndex].id,
        manager_id: approved || rejected || cancelled ? manager.id : null,
        title,
        description: `${title} for ${projectName}.`,
        start_date: date(startDate),
        end_date: date(endDate),
        capacity_percent: capacityPercent,
        priority,
        status,
        reject_reason: rejected ? 'BA has conflicting priority work in this period.' : null,
        cancel_reason: cancelled ? 'Project scope changed before kickoff.' : null,
        manager_comment: approved ? 'Approved in seed data.' : null,
        approved_at: approved ? date(startDate) : null,
        rejected_at: rejected ? date('2026-06-01') : null,
        cancelled_at: cancelled ? date('2026-05-10') : null
      }
    });
    createdBookings.push(booking);
  }

  const inboxFocusedBookings = [
    {
      ba_id: bas[4].id,
      project_id: (projectByName.get('Payment Refund Flow') ?? projects[0]).id,
      requester_id: pmUsers[4].id,
      title: 'Payment Refund Flow',
      description: 'Portal request for failed refunds and validation updates.',
      notes: 'Requested BA: Bui Phuong Thao',
      start_date: date('2026-06-01'),
      end_date: date('2026-06-05'),
      capacity_percent: 100,
      priority: BookingPriority.URGENT,
      status: BookingStatus.PENDING
    },
    {
      ba_id: null,
      project_id: (projectByName.get('CRM Revamp') ?? projects[0]).id,
      requester_id: pmUsers[1].id,
      title: 'CRM Revamp',
      description: 'Open request for dependency mapping and BA assignment.',
      notes: 'BA not assigned yet.',
      start_date: date('2026-06-03'),
      end_date: date('2026-06-06'),
      capacity_percent: 50,
      priority: BookingPriority.MEDIUM,
      status: BookingStatus.PENDING
    },
    {
      ba_id: null,
      project_id: (projectByName.get('Mobile Onboarding') ?? projects[0]).id,
      requester_id: pmUsers[3].id,
      title: 'Mobile Onboarding',
      description: 'Open request for onboarding workstream alignment.',
      notes: '[VERIFY] Needs manager verification before BA assignment.',
      start_date: date('2026-06-05'),
      end_date: date('2026-06-12'),
      capacity_percent: 100,
      priority: BookingPriority.MEDIUM,
      status: BookingStatus.PENDING
    },
    {
      ba_id: bas[3].id,
      project_id: (projectByName.get('CRM Revamp') ?? projects[0]).id,
      requester_id: pmUsers[0].id,
      title: 'Reporting Portal Upgrade',
      description: 'Specific BA request for portal reporting enhancements.',
      notes: 'Requested BA: Le Dang Khoa',
      start_date: date('2026-06-07'),
      end_date: date('2026-06-11'),
      capacity_percent: 50,
      priority: BookingPriority.MEDIUM,
      status: BookingStatus.PENDING
    },
    {
      ba_id: bas[9].id,
      project_id: (projectByName.get('BI Dashboard') ?? projects[0]).id,
      requester_id: pmUsers[2].id,
      title: 'Data Warehouse Redesign',
      description: 'Specific BA request for reporting model redesign.',
      notes: 'Requested BA: Nguyen Mai Linh',
      start_date: date('2026-06-08'),
      end_date: date('2026-06-13'),
      capacity_percent: 50,
      priority: BookingPriority.HIGH,
      status: BookingStatus.PENDING
    },
    {
      ba_id: null,
      project_id: (projectByName.get('BI Dashboard') ?? projects[0]).id,
      requester_id: pmUsers[2].id,
      title: 'Analytics Dashboard',
      description: 'Open request for analytics dashboard discovery.',
      notes: 'Needs BA assignment.',
      start_date: date('2026-06-09'),
      end_date: date('2026-06-13'),
      capacity_percent: 50,
      priority: BookingPriority.HIGH,
      status: BookingStatus.PENDING
    }
  ] as const;

  for (const input of inboxFocusedBookings) {
    const booking = await prisma.booking.create({
      data: {
        ...input,
        manager_id: null
      }
    });
    createdBookings.push(booking);
  }

  for (const [index, content] of [
    'Strong stakeholder facilitation, good fit for discovery-heavy work.',
    'Prefers clear acceptance criteria before sprint planning.',
    'Recently mentored junior BA on BPMN modeling.',
    'Watch workload near quarter end due reporting commitments.',
    'Good candidate for API-heavy projects.'
  ].entries()) {
    await prisma.privateNote.create({
      data: {
        ba_id: bas[index].id,
        content,
        created_by: manager.id
      }
    });
  }

  await prisma.notification.createMany({
    data: [
      {
        recipient_id: manager.id,
        type: 'BOOKING_REQUEST_CREATED',
        title: 'New booking request',
        message: 'A pending request needs review.',
        related_entity_type: 'Booking',
        related_entity_id: createdBookings[1].id
      },
      {
        recipient_id: pmUsers[0].id,
        type: 'BOOKING_APPROVED',
        title: 'Booking approved',
        message: 'Your request for Payment Refund Flow was approved.',
        related_entity_type: 'Booking',
        related_entity_id: createdBookings[0].id
      },
      {
        recipient_id: pmUsers[3].id,
        type: 'BOOKING_REJECTED',
        title: 'Booking rejected',
        message: 'Your CRM request was rejected with a manager reason.',
        related_entity_type: 'Booking',
        related_entity_id: createdBookings[9].id
      }
    ]
  });

  await prisma.auditLog.create({
    data: {
      actor_id: manager.id,
      action: 'SEED_DATABASE',
      target_type: 'Database',
      target_id: manager.id,
      result: 'SUCCESS',
      new_value: {
        users: 21,
        ba_profiles: 15,
        bookings: createdBookings.length
      }
    }
  });

  console.log('Seed completed', {
    users: 21,
    baProfiles: bas.length,
    pmPoUsers: pmUsers.length,
    projects: projects.length,
    tags: tags.length,
    bookings: createdBookings.length
  });
}

main()
  .catch((error) => {
    console.error(error);
    process.exitCode = 1;
  })
  .finally(async () => {
    await prisma.$disconnect();
  });

File: apps/api/cmd/api/seed.go
package main

import (
	"context"
	"fmt"
	"log"
	"net/url"
	"os"
	"strings"
	"time"

	"github.com/google/uuid"
	"golang.org/x/crypto/bcrypt"
)

func mainSeed() error {
	db, err := OpenDBFromEnv()
	if err != nil {
		return err
	}
	defer db.Close()
	ctx := context.Background()

	if err := resetDatabase(ctx, db); err != nil {
		return err
	}

	managerPasswordHash, _ := bcrypt.GenerateFromPassword([]byte("Manager@123"), 10)
	pmPasswordHash, _ := bcrypt.GenerateFromPassword([]byte("Pmpo@123"), 10)
	baPasswordHash, _ := bcrypt.GenerateFromPassword([]byte("Ba@123"), 10)
	adminPasswordHash, _ := bcrypt.GenerateFromPassword([]byte("Admin@123"), 10)

	managerID, err := insertUser(ctx, db, UserSeed{FullName: "Mai Lan Anh", Email: "manager@ba-bazaar.local", Role: "BA_MANAGER", PasswordHash: string(managerPasswordHash), AvatarURL: pravatar(11)})
	if err != nil {
		return err
	}
	if _, err := insertUser(ctx, db, UserSeed{FullName: "Bao Tri Admin", Email: "admin@ba-bazaar.local", Role: "ADMIN", PasswordHash: string(adminPasswordHash), AvatarURL: pravatar(12)}); err != nil {
		return err
	}
	hasITAdminRole, err := enumValueExists(ctx, db, "UserRole", "IT_ADMIN")
	if err != nil {
		return err
	}
	if hasITAdminRole {
		itAdminPasswordHash, _ := bcrypt.GenerateFromPassword([]byte("ItAdmin@123"), 10)
		if _, err := insertUser(ctx, db, UserSeed{FullName: "Bao Tri IT Admin", Email: "it-admin@ba-bazaar.local", Role: "IT_ADMIN", PasswordHash: string(itAdminPasswordHash), AvatarURL: pravatar(13)}); err != nil {
			return err
		}
	} else {
		log.Printf("seed: skipped IT_ADMIN user because local database enum UserRole does not include IT_ADMIN")
	}

	pmNames := []string{"Minh Tran", "Hoa Nguyen", "Quang Pham", "Linh Do", "Khanh Vo"}
	pmIDs := make([]string, 0, len(pmNames))
	for i, name := range pmNames {
		id, err := insertUser(ctx, db, UserSeed{FullName: name, Email: fmt.Sprintf("pm%d@ba-bazaar.local", i+1), Role: "PM_PO", PasswordHash: string(pmPasswordHash), AvatarURL: pravatar(21 + i)})
		if err != nil {
			return err
		}
		pmIDs = append(pmIDs, id)
	}

	tagIDs := map[string]string{}
	for _, item := range []struct{ Name, Group string }{
		{"Fintech", "DOMAIN"},
		{"E-commerce", "DOMAIN"},
		{"Logistics", "DOMAIN"},
		{"HR Tech", "DOMAIN"},
		{"CRM", "DOMAIN"},
		{"BPMN", "ANALYSIS_SKILL"},
		{"User Story Mapping", "ANALYSIS_SKILL"},
		{"Data Analysis", "ANALYSIS_SKILL"},
		{"API Specification", "ANALYSIS_SKILL"},
		{"Stakeholder Workshop", "ANALYSIS_SKILL"},
	} {
		id := uuid.NewString()
		_, err := db.Pool.Exec(ctx, `insert into skill_tags (id, name, "group", status, created_at, updated_at) values ($1, $2, $3, 'ACTIVE', now(), now())`, id, item.Name, item.Group)
		if err != nil {
			return err
		}
		tagIDs[item.Name] = id
	}

	projectIDs := map[string]string{}
	for _, item := range []struct{ Name, Color, Description string }{
		{"Payment Refund Flow", "#2563EB", "Refund and reconciliation workflow"},
		{"Mobile Onboarding", "#16A34A", "Digital onboarding for mobile users"},
		{"CRM Revamp", "#7C3AED", "Internal CRM modernization"},
		{"Logistics Tracking", "#F97316", "Shipment tracking and exception flows"},
		{"BI Dashboard", "#0F766E", "Executive utilization and delivery dashboard"},
		{"HR Approval Workflow", "#DB2777", "People operation request approvals"},
	} {
		id := uuid.NewString()
		_, err := db.Pool.Exec(ctx, `insert into projects (id, name, color, description, created_at, updated_at) values ($1, $2, $3, $4, now(), now())`, id, item.Name, item.Color, item.Description)
		if err != nil {
			return err
		}
		projectIDs[item.Name] = id
	}
	projectIDs["Internal Portal"] = projectIDs["Payment Refund Flow"]

	baInputs := []struct {
		Name       string
		Level      string
		Status     string
		TagNames   []string
		Reason     *string
		StatusDate *time.Time
	}{
		{"Pham Ngoc Chi", "SENIOR", "ACTIVE", []string{"Fintech", "BPMN", "API Specification"}, nil, nil},
		{"Do Anh Dung", "MIDDLE", "ACTIVE", []string{"Logistics", "Data Analysis"}, nil, nil},
		{"Nguyen Bao An", "JUNIOR", "ACTIVE", []string{"HR Tech", "User Story Mapping"}, nil, nil},
		{"Le Dang Khoa", "MIDDLE", "ACTIVE", []string{"CRM", "Stakeholder Workshop"}, nil, nil},
		{"Bui Phuong Thao", "SENIOR", "ACTIVE", []string{"CRM", "Data Analysis"}, nil, nil},
		{"Hoang Minh Chau", "LEAD", "ACTIVE", []string{"Fintech", "CRM", "BPMN"}, nil, nil},
		{"Tran Gia Huy", "MIDDLE", "ACTIVE", []string{"E-commerce", "API Specification"}, nil, nil},
		{"Vo Thanh Tam", "SENIOR", "ACTIVE", []string{"Logistics", "BPMN"}, nil, nil},
		{"Dang Thu Ha", "JUNIOR", "ACTIVE", []string{"CRM", "User Story Mapping"}, nil, nil},
		{"Nguyen Mai Linh", "MIDDLE", "ACTIVE", []string{"Fintech", "Data Analysis"}, nil, nil},
		{"Pham Quoc Bao", "SENIOR", "ACTIVE", []string{"E-commerce", "Stakeholder Workshop"}, nil, nil},
		{"Do Minh Tue", "MIDDLE", "ACTIVE", []string{"HR Tech", "BPMN"}, nil, nil},
		{"Le Hoai Nam", "LEAD", "ON_LEAVE", []string{"CRM", "API Specification"}, stringPtr("Temporary leave"), timePtr(dateOnly("2026-05-20"))},
		{"Vu Nhat Vy", "SENIOR", "ON_LEAVE", []string{"Fintech", "Data Analysis"}, stringPtr("Temporary leave"), timePtr(dateOnly("2026-05-20"))},
		{"Tran Thai Son", "MIDDLE", "RESIGNED", []string{"Logistics", "BPMN"}, stringPtr("Historical profile retained"), timePtr(dateOnly("2026-05-20"))},
	}
	baIDs := make([]string, 0, len(baInputs))
	for i, item := range baInputs {
		userID, err := insertUser(ctx, db, UserSeed{FullName: item.Name, Email: fmt.Sprintf("ba%d@ba-bazaar.local", i+1), Role: "BA", PasswordHash: string(baPasswordHash), AvatarURL: pravatar(41 + i)})
		if err != nil {
			return err
		}
		baID := uuid.NewString()
		joinedDate := fmt.Sprintf("202%d-0%d-15", minInt(i%5, 5), (i%8)+1)
		_, err = db.Pool.Exec(ctx, `insert into ba_profiles (id, user_id, full_name, email, phone, level, joined_date, avatar_url, status, status_reason, status_changed_at, created_at, updated_at, version) values ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,now(),now(),1)`, baID, userID, item.Name, fmt.Sprintf("ba%d@ba-bazaar.local", i+1), fmt.Sprintf("090%07d", i+1), item.Level, dateOnly(joinedDate), pravatar(41+i), item.Status, item.Reason, item.StatusDate)
		if err != nil {
			return err
		}
		for _, tagName := range item.TagNames {
			_, err = db.Pool.Exec(ctx, `insert into ba_skill_tags (id, ba_id, tag_id, assigned_by, assigned_at) values ($1, $2, $3, $4, now())`, uuid.NewString(), baID, tagIDs[tagName], managerID)
			if err != nil {
				return err
			}
		}
		baIDs = append(baIDs, baID)
	}

	createdBookings, err := seedBookings(ctx, db, managerID, pmIDs, baIDs, projectIDs)
	if err != nil {
		return err
	}
	if err := seedNotesNotificationsAudit(ctx, db, managerID, pmIDs, baIDs, createdBookings); err != nil {
		return err
	}

	hasAiFeatureFlags, err := tableExists(ctx, db, "ai_feature_flags")
	if err != nil {
		return err
	}
	hasAiSettings, err := tableExists(ctx, db, "ai_settings")
	if err != nil {
		return err
	}

	if hasAiFeatureFlags {
		flags := []struct {
			Key, Name, Desc string
			Enabled         bool
		}{
			{"ai_suggest_ba_enabled", "AI Suggest BA", "Gate for Suggest BA feature", true},
			{"ai_prd_skill_extraction_enabled", "PRD Skill Extraction", "Gate for extracting skills from PRD", true},
			{"ai_pmpo_chatbot_enabled", "PM/PO Chatbot", "Gate for future PM/PO chatbot", true},
			{"ai_ba_manager_chatbot_enabled", "BA Manager Chatbot", "Gate for future BA Manager chatbot", true},
			{"ai_dashboard_summary_enabled", "Dashboard Summary", "Gate for LLM dashboard summary", true},
			{"ai_report_summary_enabled", "Report Summary", "Gate for LLM report summary", true},
			{"ai_observability_enabled", "AI Observability logging", "Master switch for logging AI session data", true},
		}
		for _, f := range flags {
			_, err = db.Pool.Exec(ctx, `
				insert into ai_feature_flags (id, key, name, description, enabled, environment, updated_at)
				values ($1, $2, $3, $4, $5, 'all', now())
			`, uuid.NewString(), f.Key, f.Name, f.Desc, f.Enabled)
			if err != nil {
				return fmt.Errorf("seed flags: %w", err)
			}
		}
	}

	if hasAiSettings {
		settings := []struct {
			Key, Value, Desc string
		}{
			{"ai_model_name", "deepseek-chat", "Primary LLM model name used for AI features"},
			{"ai_temperature", "0.2", "Temperature parameter for AI generation"},
			{"ai_prompt_version", "v1.0.0", "Active prompt version tag for logging"},
		}
		for _, s := range settings {
			_, err = db.Pool.Exec(ctx, `
				insert into ai_settings (id, key, value, description, updated_at)
				values ($1, $2, $3, $4, now())
			`, uuid.NewString(), s.Key, s.Value, s.Desc)
			if err != nil {
				return fmt.Errorf("seed settings: %w", err)
			}
		}
	}

	if !hasAiFeatureFlags || !hasAiSettings {
		log.Printf("seed: skipped AI defaults because local database is missing AI observability tables")
	}

	log.Printf("seeded users=%d ba_profiles=%d projects=%d tags=%d bookings=%d", 22, len(baIDs), len(projectIDs), len(tagIDs), len(createdBookings))
	return nil
}

type UserSeed struct {
	FullName     string
	Email        string
	Role         string
	PasswordHash string
	AvatarURL    string
}

func insertUser(ctx context.Context, db *DB, input UserSeed) (string, error) {
	id := uuid.NewString()
	_, err := db.Pool.Exec(ctx, `insert into users (id, full_name, email, role, password_hash, avatar_url, created_at, updated_at) values ($1,$2,$3,$4,$5,$6,now(),now())`, id, input.FullName, input.Email, input.Role, input.PasswordHash, input.AvatarURL)
	return id, err
}

func resetDatabase(ctx context.Context, db *DB) error {
	tables := []string{
		"audit_logs", "notifications", "private_notes", "bookings",
		"ba_skill_tags", "skill_tags", "projects", "ba_profiles",
		"refresh_tokens", "users",
	}

	optionalTables := []string{
		"ai_settings", "ai_feature_flags", "ai_feedback", "ai_errors",
		"ai_tool_calls", "ai_extractions", "ai_messages", "ai_sessions",
	}

	for _, table := range optionalTables {
		exists, err := tableExists(ctx, db, table)
		if err != nil {
			return err
		}
		if !exists {
			continue
		}
		if _, err := db.Pool.Exec(ctx, "delete from "+table); err != nil {
			return err
		}
	}

	for _, table := range tables {
		if _, err := db.Pool.Exec(ctx, "delete from "+table); err != nil {
			return err
		}
	}
	return nil
}

func pravatar(imageID int) string { return fmt.Sprintf("https://i.pravatar.cc/300?img=%d", imageID) }
func dateOnly(value string) time.Time {
	t, _ := time.Parse("2006-01-02", value)
	return t.UTC()
}
func stringPtr(value string) *string { return &value }
func timePtr(value time.Time) *time.Time { return &value }
func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func tableExists(ctx context.Context, db *DB, tableName string) (bool, error) {
	var exists bool
	err := db.Pool.QueryRow(ctx, `
		select exists (
			select 1
			from information_schema.tables
			where table_schema = 'public' and table_name = $1
		)
	`, tableName).Scan(&exists)
	return exists, err
}

func enumValueExists(ctx context.Context, db *DB, enumName string, enumValue string) (bool, error) {
	var exists bool
	err := db.Pool.QueryRow(ctx, `
		select exists (
			select 1
			from pg_type t
			join pg_enum e on e.enumtypid = t.oid
			where t.typname = $1 and e.enumlabel = $2
		)
	`, enumName, enumValue).Scan(&exists)
	return exists, err
}

func init() {
	if len(os.Args) > 1 && os.Args[1] == "seed" {
		if strings.TrimSpace(os.Getenv("DATABASE_URL")) == "" {
			os.Setenv("DATABASE_URL", resolveLocalDatabaseURL())
		}
		if err := mainSeed(); err != nil {
			log.Fatal(err)
		}
		os.Exit(0)
	}
}

func normalizeDSN(dsn string) string {
	parsed, err := url.Parse(dsn)
	if err != nil {
		return dsn
	}
	q := parsed.Query()
	q.Del("schema")
	parsed.RawQuery = q.Encode()
	return parsed.String()
}
